### PR TITLE
RND203-363: Add mutexes

### PR DESCRIPTION
**Pull request by nzulinski-mobica**

### CODE REVIEW

**Comment by nzulinski-mobica** [comment]
There still is a problem with mutexes 
         * Put it to sleep for now.
	 */
	//ATH_LOCK(sc);
	ath_power_setpower(sc, HAL_PM_FULL_SLEEP, 1);
	//ATH_UNLOCK(sc);
Because it's in neverending sleep and waiting for other thread to end

File: sys/dev/ath/if_ath_usb.c [hunk 1]
```diff
@@ -462,26 +462,32 @@ ath_usb_attach(device_t self)
...
 #endif
 	error = usbd_transfer_setup(uaa->device, &uaa->info.bIfaceIndex, usc->sc_xfer, ath_if_config,
-								ath_N_XFER, usc, &sc->sc_mtx);
+								ATH_N_XFER, usc, &sc->sc_usb_mtx);
 	if (error) {
 		device_printf(sc->sc_dev,
 					  "could not allocate USB transfers, err=%s\n",
 					  usbd_errstr(error));
-		mtx_destroy(&sc->sc_mtx);
+		ATH_TXSTATUS_LOCK_DESTROY(sc);
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/ath/if_ath_usb.c:485]: I think we should also destroy the mutexes in the `bad` label.

File: sys/dev/ath/if_athvar.h [hunk 2]
```diff
@@ -607,6 +607,8 @@ struct ath_softc {
 	char			sc_tx_mtx_name[32];
 	struct mtx		sc_tx_ic_mtx;	/* TX queue mutex */
 	char			sc_tx_ic_mtx_name[32];
+	struct mtx		sc_usb_mtx;	/* USB access mutex */
```
Comment (Mikolaj-Filar-Mobica) [inline sys/dev/ath/if_athvar.h:610]: `ath_softc *sc = device_get_softc(self);` -> Can we modify the `ath_softc?` structure? I don't know this mechanism.

File: sys/dev/ath/if_ath.c [hunk 3]
```diff
@@ -1383,9 +1383,9 @@ ath_attach(u_int16_t venid, u_int16_t devid, struct ath_softc *sc)
 	/*
 	 * Put it to sleep for now.
 	 */
-	ATH_LOCK(sc);
+	//ATH_LOCK(sc);
 	ath_power_setpower(sc, HAL_PM_FULL_SLEEP, 1);
-	ATH_UNLOCK(sc);
+	//ATH_UNLOCK(sc);
```
Comment (Rafal-Litwinski-Mobica) [inline sys/dev/ath/if_ath.c:1391]: Mutex lock is required for other ATH PCIe devices.

File: sys/dev/ath/if_athvar.h (hunk 2 above)
Comment (Rafal-Litwinski-Mobica) [reply sys/dev/ath/if_athvar.h:610]: device_get_softc returns only a pointer do sc struct:
return (dev->dev_sc);
So it's not a problem if we modify the ath_softc structure.

File: sys/dev/ath/if_ath_usb.c (hunk 1 above)
Comment (Rafal-Litwinski-Mobica) [reply sys/dev/ath/if_ath_usb.c:485]: Done

**Review by Mikolaj-Filar-Mobica — APPROVED** [review]

File: sys/dev/ath/if_ath.c (hunk 3 above)
Comment (Rafal-Litwinski-Mobica) [reply sys/dev/ath/if_ath.c:1391]: TODO comment added, new task created for this issue.

**Review by Rafal-Litwinski-Mobica — APPROVED** [review]

**Review by Mikolaj-Filar-Mobica — APPROVED** [review]

**Review by Mikolaj-Filar-Mobica — APPROVED** [review]

**Pull request merged**